docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+
+If we add another property; invertibility, we could technically _undo_ upgrade steps and as such, if an error is inserted in the process, we could revert the step, make the corrections, then apply the step again.
+
+### **RFC: should invertibility be a requirement or ignored.**
+
+## Step Definition
+
+In the CLI, the `upgrade/steps/` directory will be kept for step scripts. It has the following structure (example):
+
+    angular-cli/
+      upgrade/
+        index.ts                    <-- the Upgrade api (see below).
+        steps/
+          angular2/                 <-- or other npm packages.
+            0001/                   <-- incrementing number.
+              index.ts              <-- the script which will be executed to move up.
+              XYZ.ts                <-- some utility used by a step.
+
+The `index.ts` of the step exports the following symbols:
+
+* `upgrade(project: AngularProject): Observable<Change>`; upgrade from the current version to the `toVersion`. Returns a stream of changes.
+* `downgrade(project: AngularProject): Observable<Change>`; downgrade from the `toVersion` to the version of this script. Returns a stream of changes.
+* `fromVersion: SemVersion | SemVersion[]`; either a version string or an array of all version strings representing the versions this package must be before `up()` is called.
+* `toVersion: SemVersion`; a version string representing the current version when `up()` succeeds.
+
+### Rationales
+An incrementing number is used instead of the from version because some versions might not be a 1-to-1 relationship between two versions. For example, version `beta-12` and `beta-13` might be both no-op, and the script upgrading to `beta-14` might have a `fromVersion` of `[beta-12, beta-13]`. Using a binary search, the lookup time is not a factor.
+
+Using 4 digits allow us to sort the folder alphabetically while preserving order, and is more than enough; at 1 release per day this would take 27 years.
+
+
+## Refactoring
+Everything in a project is changeable during an upgrade; from templates, to the angular-cli config or npm packages, or TypeScript or Javascript code.
+
+In all those cases, the input of the upgrade function is a project (which includes configuration, packages, root tree, build files, ...) and the output will be a list of changes. These changes might be applied or not, but the upgrade function should not have any side effects.
+
+### Code
+An AST-to-AST transformation step will be necessary to perform complex refactoring. The internal `upgrade` library can be used to build a transformation pipeline that will be applied to every source files in the Angular app.
+
+The pipeline is a stream of `ts.SourceFile`, and should receive a list of changes to apply to those sources.
```

Comment:
What about html/css changes? We can tell users to do it manually, but we need to do it otherwise code might compile but be utterly broken.